### ISSUE TITLE
(txn|session).run sometimes does not return 

### ISSUE THREAD

**Issue by cipacda** [neo4j/neo4j-javascript-driver#191]
- Neo4j version: 3.0.6
- Operating system: Windows 8 and 10
- API/Driver: Neo4j Javascript driver - neo4j-driver@1.1.0-M04

##Steps to reproduce - Behavior##
This is not easy to reproduce as it does not have an exact replication and it doesn't always happen, but I will try to be as explicit as possible, please let me know if you need more clarification.

I am firing a few requests at the same time to the database and some of them (not all) are failing with a `Neo.TransientError.Transaction.DeadlockDetected` and I have a system in place to wait for some time (1 second in tests) and then rollback the transaction (in this case I can't rollback because the rollback would have already caused the rollback, but I am doing that asynchronously so it should not affect my process at all). Once the transaction is rolled back, I am creating a new transaction and am retrying all the queries again.

In the second run, the first query always seems to succeed but sometimes some of the following ones are timing out (I have a timeout on the promise running the query - the driver does not return anything back). This is my code running the query:

````
return new Promise((resolve, reject) => {
        txn.run(queryString, params)
            .then(results => {
                    resolve(results);
                })
            .catch((e) => {
                reject(e);
            });
    }).timeout(50000, 'Timeout encountered while running query ' + queryString);
````

Normally all queries end in at most 30ms, the Neo4J instance running on the same machine but in this case, I am getting nothing back from Neo4J in 50 seconds, which is a very long time for a request. 

Please note that in some cases this second run after the deadlock succeeds and there is no issue.

Not sure if related, but I am getting a timeout sometimes when I am running the queries directly in a session without getting a deadlock before, not sure is related, but the same code as above is used, except that txn refers to a session then.

##Expected behavior##
I would expect that the driver always returns something and if encounters an error of some kind it should report on it and not hang.

##Additional comments##
As part of my 2 days of investigation on this, I have ended up adding logging in the neo4j-driver library to see what actually happens in the driver and you can see further down important events I logged. I have more data if needed, but I can't share the exact log for confidentiality reasons. Anyhow I am happy to reply to any questions you might have or provide extra info if you need.

1. The transaction is in the ACTIVE state when running the query for which I am getting no reply.
2. Within 2ms, the run function inside the transaction is completed without error (in transaction.js), the Result model is created (in result.js)
3. Nothing happens on **the connection** in which the transaction is running for 50 seconds, after which running transactions are stopped with `Neo.TransientError.Transaction.LockClientStopped` because my application is closing the sessions after the timeout of 50 seconds.
4. I have logs in the following places which are not reached at all for the problematic query:  
\- onNext and onCompleted inside _createPromise in result.js
\- onCompletedWrapper inside subscribe in result.js
\- on the catch branch inside the run for ACTIVE state (actuall for all states) in transaction.js
\- inside onCompleted and onError inside _TransactionStreamObserver in transaction.js
\- inside _handleFatalError and _handleMessage in connector.js
\- some others I may have missed

As other queries are quickly reaching the onNext/onComplete for the created Result I am pretty certain that in this case the driver gets stuck somehow (probably even the db is not replying).

Could you please investigate and let me know if I can avoid this in any way or how long it will take for you to fix it? It is a blocking issue for our current project and it delays our beta launch as we can't have asynchronous calls.

Your help is much appreciated

**Comment by lutovich** [comment]
Hi @cipacda,

Could you please check neo4j database logs for periods when this issue occurs? Those are `neo4j.log` and `debug.log`.

**Comment by cipacda** [comment]
There is no error reported around that time (19.01.2017 - 14:55 - this is when timeout ocurred after deadlock) but I have attached the log id helpful.

It can be seen that some threads were blocked at some time, which looks related.

[querytimeout-debug.txt](https://github.com/neo4j/neo4j-javascript-driver/files/717009/querytimeout-debug.txt)

**Comment by lutovich** [comment]
Hi @cipacda,

Sorry for the late reply. It looks like there is a rather large GC pause around 19.01.2017 - 14:55:

```
2017-01-19 14:54:56.389+0000 WARN  [o.n.k.i.c.MonitorGc] GC Monitor: Application threads blocked for 18470ms.
```

This could cause queries to hang because database was busy with garbage collection. `LockClientStopped` is basically `TransactionTerminatedException` which was detected while trying to grab a lock. So this means database was actually trying to execute the query and take all needed locks. Seems like driver was just waiting for response on a valid connection.

Do you mind sharing memory settings (page cache and heap) and data sizes (`ls -lah` in the `graph.db` directory)? It is worth enabling GC logging as well as described here: https://neo4j.com/developer/kb/enabling-gc-logging/ so we could try to correlate query hanging with long GC pauses.

Meanwhile I'll try to reproduce deadlocks in a test application with different database memory settings.

**Comment by lutovich** [comment]
@cipacda I've tried to reproduce the problem with this gist: https://gist.github.com/lutovich/a6942a8856ea645b4ba7d781bac14744.

It executes queries like:

```
MATCH (u1:User{uid: {user1}}), (u2:User{uid: {user2}}), (u3:User{uid: {user3}})
CREATE (u1)-[l1:LIKE {created: {created1}}]->(u2), (u2)-[l2:LIKE {created: {created2}}]->(u3)
RETURN (u1)-[l1]->(u2)-[l2]->(u3), range(0, 10000)
```

for random user ids taken from a list of 50. Set of users is limited and each query locks 3 nodes. Deadlocks occur fairly often. `range(0, 10000)` is present only to force database to create more objects. Queries are executed asynchronously with at most 12 operations at a time. Database is 3.0.6 with 20mb of page cache and 500mb or heap memory.

This is the output or `test.js` script:

```
Users created
Index created
Stats: {"queriesExecuted":1000,"deadlocksDetected":2,"maxQueryExecutionTime":171}
Stats: {"queriesExecuted":2000,"deadlocksDetected":2,"maxQueryExecutionTime":171}
Stats: {"queriesExecuted":3000,"deadlocksDetected":5,"maxQueryExecutionTime":186}
Stats: {"queriesExecuted":4000,"deadlocksDetected":7,"maxQueryExecutionTime":186}
Stats: {"queriesExecuted":5000,"deadlocksDetected":11,"maxQueryExecutionTime":196}
Stats: {"queriesExecuted":6000,"deadlocksDetected":14,"maxQueryExecutionTime":196}
Stats: {"queriesExecuted":7000,"deadlocksDetected":15,"maxQueryExecutionTime":196}
....
Stats: {"queriesExecuted":97000,"deadlocksDetected":131,"maxQueryExecutionTime":234}
Stats: {"queriesExecuted":98000,"deadlocksDetected":131,"maxQueryExecutionTime":234}
Stats: {"queriesExecuted":99000,"deadlocksDetected":133,"maxQueryExecutionTime":234}
Stats: {"queriesExecuted":99000,"deadlocksDetected":134,"maxQueryExecutionTime":234}
```

so this script does not expose issue you are seeing. Deadlocks happen but queries do not slow down much.

Could you maybe share couple queries that observe the problem? I'll then try to generate test data and run similar script with your queries.

Thanks!

**Comment by lutovich** [comment]
Hi @cipacda,

Is this problem still bugging you? Did maybe have a change to look at the script I used to reproduce the problem?

You mentioned some facility that helps you with retries. We've introduced couple new API functions in 1.2.0-rc1 release that do retries with exponential backoff. They might be useful to you and your feedback would be super useful. See last entry in the [changelog](https://github.com/neo4j/neo4j-javascript-driver/wiki/1.2-changelog#120-rc1) for more details.

Thanks.
